### PR TITLE
V1.2 rust

### PR DESCRIPTION
**Pull request by ErisMik**

### CODE REVIEW

File: .github/workflows/rust-demos.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    - name: Test
+      run: cargo run --release -- --show_audio_devices
+
+  build-self-hosted:
+    runs-on: ${{ matrix.machine }}
+
+    strategy:
+      matrix:
+        machine: [rpi3-32, rpi3-64, rpi4-32, rpi4-64, jetson]
```
Comment (ksyeo1010) [inline .github/workflows/rust-demos.yml:62]: we don't do beaglebone?

File: binding/rust/README.md [hunk 2]
```diff
@@ -49,10 +49,9 @@ recorder.start()?
 Get the pcm frames by calling the read function:
 
 ```rust
-loop {
-    let mut pcm_frame_buffer: [i16; FRAME_LENGTH] = [0; FRAME_LENGTH];
-    recorder.read(&mut pcm_frame_buffer)?;
-    // do something with pcm
+while recorder.is_recording() {
+    let frame = recorder.read()?;
+    // do something with pcm frame
```
Comment (ksyeo1010) [inline binding/rust/README.md:54]:
```suggestion
    // do something with frame
```

File: binding/rust/src/pvrecorder.rs [hunk 3]
```diff
@@ -154,20 +149,18 @@ impl RecorderBuilder {
 
     pub fn init(&self) -> Result<Recorder, RecorderError> {
```
Comment (laves) [inline binding/rust/src/pvrecorder.rs:150]: In all other SDKs, we're calling the classes PvRecorder and PvRecorderError/Exception - can we follow that pattern here?

File: binding/rust/README.md [hunk 4]
```diff
@@ -8,9 +8,9 @@ A cross platform audio recorder that captures single-channel audio at a sample r
 
 ## Compatibility
```
Comment (laves) [inline binding/rust/README.md:9]: It's pedantic, but all the other titles are `PvRecorder Binding for <SDK>`

**Review by laves — CHANGES_REQUESTED** [review]

File: binding/rust/src/pvrecorder.rs [hunk 5]
```diff
@@ -93,14 +102,12 @@ impl std::error::Error for RecorderError {}
 
 const DEFAULT_DEVICE_INDEX: i32 = -1;
 const DEFAULT_FRAME_LENGTH: i32 = 512;
```
[the remark this answers is not in the document]
Comment (ErisMik) [reply binding/rust/src/pvrecorder.rs:104]: I left it in so we can satisfy the `Default` trait for the binding, without it getting the `available` devices is a bit awkward. But I changed the main builder workflow to require an explicit frame length

File: demo/rust/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+It provides developers with access to the input stream of an audio device, which is divided into data frames of a given size.
+
+## Requirements
+
+- Rust 1.54+
+
+## Compatibility
+
+- Linux (x86_64)
+- macOS (x86_64)
```
Comment (laves) [inline demo/rust/README.md:17]: arm64

File: binding/rust/README.md [hunk 7]
```diff
@@ -32,27 +30,26 @@ pv_recorder = "*"
...
 ```
 
 To start recording initialize an instance using the builder and run `start`:
 
 ```rust
-use pv_recorder::RecorderBuilder;
+use pv_recorder::PvRecorderBuilder;
 
-let recorder = RecorderBuilder::new().init()?;
+let recorder = PvRecorderBuilder::new(512).init()?;
```
Comment (laves) [inline binding/rust/README.md:43]: does rust have argument labels, i.e. `PvRecorderBuilder::new(frameLength: 512).init()?;` ?

If not, you should mention it in a comment or in the sentence above

**Review by laves — APPROVED** [review]

File: binding/rust/README.md (hunk 7 above)
Comment (ErisMik) [reply binding/rust/README.md:43]: No argument labels, I'll set a variable

**Comment by ErisMik** [comment]
Released!

**Pull request merged**